### PR TITLE
Reworking CDSAPI.jl to be more Julia-centric

### PR DESCRIPTION
**Pull request by natgeo-wong**
* Removed py2ju -> people who are using CDSAPI.jl should be more or less familiar with Julia Dictionaries
* Added some logging statements to `retrieve()` to mimic the Python CDSAPI retrieval logs
* Updated Project.toml to reflect all changes, and allow more flexibility in Julia version
* Bump version to 0.2.0 because of breaking changes/syntax

### CODE REVIEW

File: src/CDSAPI.jl [hunk 1]
```diff
@@ -1,88 +1,103 @@
...
 
 """
-    retrieve(name, params, filename)
-
-Retrieves data for `name` from the Climate Data Store
-with the specified `params` and stores it in the current
-directory as `filename`.
+    retrieve(
+        fname::AbstractString,
+        cdsdataset::AbstractString,
```
Comment (juliohm) [inline src/CDSAPI.jl:14]: I find this order confusing compared to the Python equivalent. Can we try to preserve the first argument as the name of the dataset in CDS, and the last argument as the name of the file saved?

**Comment by juliohm** [comment]
Thank you @natgeo-wong for the nice refactoring and logging. Could you please update the tests accordingly so that we have a clearer picture of the proposed API?

Appreciate if you can review the PR @LakshyaKhatri. It is also an opportunity to figure out if that issue about decryption is still present.

File: src/CDSAPI.jl (hunk 1 above)
Comment (natgeo-wong) [reply src/CDSAPI.jl]: Alright, I resorted the arguments in the newest push!

**Comment by LakshyaKhatri** [comment]
Hello @juliohm! Sorry for the delayed response. I think we should test these chages locally before merging them. Also, since this is an external pull request, Travis is turning the environment variables off to prevent modification of the encrypted files and that is why the tests are failing. 

I will first test the current changes locally and then we will be ready to merge the changes. After that, I will check what can be done on the Travis part to resolve the issue.

Also, @natgeo-wong, thank you so much for such beautiful changes. The code looks neat.

**Comment by natgeo-wong** [comment]
I revamped the set of tests, because the output is quite different. You could add your own tests to this as well?

File: src/CDSAPI.jl [hunk 2]
```diff
@@ -1,88 +1,103 @@
...
-    end
+function retrieve(
+    dataset::AbstractString,
+    params::AbstractDict,
+    fname::AbstractString,
+    keys::AbstractDict = cdskeys()
+)
 
-    apikey = string("Basic ", base64encode(creds["key"]))
+    @info "$(now()) - Welcome to the Climate Data Store"
```
Comment (LakshyaKhatri) [inline src/CDSAPI.jl:34]: Similarly `now()` is also asking for `Dates.jl` as a dependency. I am not sure these will work without dependencies or not. Please correct me if I am getting it wrong.

File: src/CDSAPI.jl [hunk 3]
```diff
@@ -1,88 +1,103 @@
...
 
+using Base64
 using HTTP
 using JSON
-using Base64
+using Logging
 
 export
-    py2ju
+        retrieve, cdskeys
```
[the remark this answers is not in the document]
Comment (natgeo-wong) [reply src/CDSAPI.jl:9]: This is my stylistic choice when exporting functions: two tab spaces instead of one

File: src/CDSAPI.jl [hunk 4]
```diff
@@ -1,88 +1,103 @@
...
+    @info """$(now()) - Downloading $(uppercase(dataset)) data
+      $(BOLD("URL:"))         $(data["location"])
+      $(BOLD("Destination:")) $(fnc)
+    """
+
+    dt1 = now()
+    HTTP.download(data["location"],fname,update_period=Inf)
+    dt2 = now()
+
+    @info "$(now()) - Downloaded $(@sprintf("%.1f",data["content_length"]/1e6)) MB in $(@sprintf("%.1f",Dates.value(dt2-dt1)/1000)) seconds (Rate: $(@sprintf("%.1f",data["content_length"]/1e3/Dates.value(dt2-dt1))) MB/s)"
```
[the remark this answers is not in the document]
Comment (natgeo-wong) [reply src/CDSAPI.jl:76]: Oh yes I forgot about this.

File: src/CDSAPI.jl (hunk 2 above)
Comment (natgeo-wong) [reply src/CDSAPI.jl:34]: I forgot about this too.  Gimme a bit, I'll make the changes.

File: src/CDSAPI.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/CDSAPI.jl:14]: Maybe rewrite to a single line `export retrieve, cdskeys` given that we only export these two names. That avoids the extra line with too much whitespace.

File: src/CDSAPI.jl [hunk 5]
```diff
@@ -1,88 +1,108 @@
...
+        keys::AbstractDict = cdskeys()
+    )
+
+Retrieves datasets from the Climate Data Store, with options specified in a Julia Dictionary and saves it into a specified file.
+
+Arguments:
+    * `fname::AbstractString` : string that contains the path and name of the file that the data is to be saved into
+    * `dataset::AbstractString` : string specifies the name of the dataset within the Climate Data Store that the `retrieve` function is attempting to retrieve data from
+    * `params::AbstractDict` : dictionary that contains the keywords that specify the properties (e.g. date, resolution, grid) of the data being retrieved
+    * `keys::AbstractDict` : dictionary that contains API Key information read from the .cdsapirc file in the home directory (optional)
```
Comment (juliohm) [inline src/CDSAPI.jl:30]: Thanks for reordering the arguments. Maybe it would make sense to reorder the docstring as well?

Comment (natgeo-wong) [reply src/CDSAPI.jl]: Done!

**Comment by LakshyaKhatri** [comment]
The current version of `cdskeys()` is showing timestamps like this:

time_stamp

---
IMO, having a timestamp won't be that much of a help. Also, for showing a timestamp we are adding an additional dependency of `Dates.jl` (which could be overkill). @juliohm, it would be helpful to have your views on this.

**Comment by juliohm** [comment]
Thanks for pinging @LakshyaKhatri , I think the dependency on Dates is fine as it is a stdlib shipped with Julia. My main concern is actually the verbosity of these messages. It is nice to have logging features, but we should consider adding a `verbose` option to do things quietly if possible. This option could default to `true`. What do you think?

File: src/CDSAPI.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliohm) [reply src/CDSAPI.jl:14]: What do you think about this one @LakshyaKhatri ? Would be ok to make it a single line?

**Comment by natgeo-wong** [comment]
The verbosity of these messages are very similar to the CDSAPI by Python, and actually I like them because I can monitor the progress of the downloads.

When I was using the initial version of CDSAPI, my datasets are extremely large and take a long time to download, so the CDSAPI script kept throwing "Request is running" every second, which I believe is even more verbose for large datasets.

Reason why I added the log for the cdskeys() is so that people are reminded that they shouldn't be continuously calling the cdskeys() function in a loop when once will do.

File: src/CDSAPI.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (LakshyaKhatri) [reply src/CDSAPI.jl:14]: Right! that would work @juliohm.

**Comment by LakshyaKhatri** [comment]
> My main concern is actually the verbosity of these messages. It is nice to have logging features, but we should consider adding a `verbose` option to do things quietly if possible. This option could default to `true`. What do you think?

@juliohm could you provide some more information. I am still linking this with the timestamps that are added by @natgeo-wong in the info logs!

File: src/CDSAPI.jl [hunk 6]
```diff
@@ -1,88 +1,116 @@
...
+        data = JSON.parse(String(data.body))
+    end
+
+    if data["state"] == "completed"
+
+        @info "$(now()) - Request is completed"
+
+        @info """$(now()) - Downloading $(uppercase(dataset)) data
+          $(BOLD("URL:"))         $(data["location"])
+          $(BOLD("Destination:")) $(fnc)
```
Comment (LakshyaKhatri) [inline src/CDSAPI.jl:76]: Also @natgeo-wong, while calling `retrieve()`, it is throwing `UndefVarError: BOLD not defined` error. Am I missing something?

Comment (natgeo-wong) [reply src/CDSAPI.jl:76]: Damn, forgot about Crayons.Box HAHA.  gimme a minute, will add the verbose option and the Crayons package.

**Comment by juliohm** [comment]
> Also, I should add that when I have more time I'm going to try and add a progress bar option to HTTP.jl for downloads, which would be cool.

That would be very cool indeed @natgeo-wong :100: 

Also, I still we should consider adding a verbose option to the download command. Sometimes users are not interested in verbosity, for example when running multiple processes in a computer cluster. I had similar issue with the Pkg manager recently and they are considering adding a verbose option there as well. Imagine multiple processes downloading different datasets.

We could just follow a pattern `verbose && @info "foo"` throughout the code. Shouldn't be that hard to add this option.

Regarding the dependency on Crayons, do we really want to add an extra dependency just for a boldface?

**Comment by natgeo-wong** [comment]
> We could just follow a pattern `verbose && @info "foo"` throughout the code. Shouldn't be that hard to add this option.

I have done this in the most recent push

> Regarding the dependency on Crayons, do we really want to add an extra dependency just for a boldface?

I could remove it haha, I just like nice formatting.

**Comment by juliohm** [comment]
Would it be ok to remove the Crayons dependency? Given that the package is very small, and that we are not using formatting all over the place, I think it would be better to remove it. I didn't check the loading time, but the less dependencies the better in the long-term.

**Comment by natgeo-wong** [comment]
Done!

**Comment by juliohm** [comment]
Awesome @natgeo-wong , do we also need an update to the README? I will try test the PR locally tomorrow :+1:

**Comment by natgeo-wong** [comment]
An update to the README would be good.  I'll work on the README over this coming week as a way to soothe myself when everything starts getting extremely crazy on Tuesday.

**Comment by juliohm** [comment]
Hey @natgeo-wong , do you think you could manage a final stretch on this PR this week? I know it is been a crazy week for you. :)

Looking forward to review the changes.

**Comment by natgeo-wong** [comment]
This week is good!  Expect updates today or tomorrow!

**Comment by natgeo-wong** [comment]
Done!

**Comment by juliohm** [comment]
I've just cloned the repo and tried the changes locally, and unfortunately I don't think we can merge the PR as is. There are many issues including the fact that tests with the ERA5 in GRIB format were erased, NetCDF was replaced with NCDatasets without a corresponding update in the Project.toml, and other source code formatting issues. Overall, it is more difficult to understand what the code is doing now compared to the master branch.

Taking a step back, this PR is basically adding a "verbose" feature, i.e. more control over the messages printed to the terminal. If you would like to start a fresh PR to add this feature, we'd be happy to review it.

**Comment by natgeo-wong** [comment]
I am sorry about the tests.  I think I mentioned early on that I am not too sure how Testing works in Julia and I'm still trying to figure it out, because I do not use it regularly.

This PR is more than just "verbosity". When I first looked at CDSAPI.jl, I was very uncertain as to how to use it because of py2ju. I've gotten rid of that, because the crux of CDSAPI.jl should be on Julia, and not on "conversion" measures from Python to Julia.  I didn't even know what the nature of the inputs were until I had a very close look at the code and realized that what py2Ju was doing was simply returning a dictionary. So I didn't even know what inputs I was supposed to be using, and that is offputting to people who are still trying out the basics of Julia.

I would argue that when one comes from Julia, one finds the master branch of CDSAPI harder to absorb because of conflicting information and priorities aimed to "converting" a python code or dictionary or information into Julia, when one should simply be using Julia code directly.  I rewrote the code to be able to do that, specifically by restricting the types of the input argument, which should already be giving people familiar with Julia a sense of what sort of input arguments are required.

I've also pulled out the retrieval of the cdskeys as a function on its own, so people don't need to keep rereading the cdskeys each time when we use loops to retrieve large datasets (over different months/years, for example).

I can probably restore some of the GRIB tests, etc. from the master branch.  I don't use those formats, so I am predisposed to rely only on NCDatasets, but I will do my best.

My key message is this: We really need to remove the dependency of py2ju, because that is just confusing.

**Comment by juliohm** [comment]
> TLDR, what I did was:
>
> * Restrict the nature of input arguments to make it clearer for Julia users

Restricting input arguments is a Julia anti-pattern, see section "Over-constraining argument types" https://www.oxinabox.net/2020/04/19/Julia-Antipatterns.html

> * Remove py2ju because we really should not be relying on python syntax in Julia code. Instead, show users how to input the arguments in Julia code.

Isn't it true that the current master branch supports Julia dicts just fine? We are only providing an additional helper function `py2ju` in case someone doesn't want to build the dict by hand. The CDS website creates Python queries for you, and converting these queries by hand can be a waste of time. What we should do is clarify in the README that Julia dicts are supported.

> * Pulling out minor functions (see below), so that they can be individually modified without disturbing the overall code.
>   a. Retrieval of cdskeys
>   b. Parsing of HTTP request

To be honest I don't see much use in a separate cdskey function, but we can review a separate PR with this improvement. Regarding the parsing of HTTP, this is the part of the code that I thought could be reformatted. What about we start two separate PRs with these changes?

**Comment by natgeo-wong** [comment]
Hi, sorry for the silence.

I will make the PRs when I have time, now is crunch time with class projects coming up.

I still believe that specifying the input types will be helpful, especially if people are coming over from Python to Julia, for new users to understand exactly how the Julia portions of the code work.  I stand by my previous statement that specifying it is necessary.

I will make separate PRs for the parts you've specified.

**Comment by juliohm** [comment]
Awesome. I will close this PR for now, and then we can start fresh PRs with individual changes when you are over your other tasks in your todo list.
